File: core/src/test/java/com/jinloes/prpilot/services/PatternKnowledgeBaseTest.java
package com.jinloes.prpilot.services;

import static org.assertj.core.api.Assertions.assertThat;

import java.io.File;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

class PatternKnowledgeBaseTest {

    @Nested
    class Load {

        @Test
        void returnsEmpty_whenNoFileExists(@TempDir File dir) {
            PatternKnowledgeBase kb = new PatternKnowledgeBase(dir);
            assertThat(kb.load("owner", "repo")).isEmpty();
        }

        @Test
        void returnsContents_afterAppend(@TempDir File dir) {
            PatternKnowledgeBase kb = new PatternKnowledgeBase(dir);
            kb.append("owner", "repo", "the comment", "the finding");
            assertThat(kb.load("owner", "repo")).isNotBlank();
        }

        @Test
        void separatesReposByFile(@TempDir File dir) {
            PatternKnowledgeBase kb = new PatternKnowledgeBase(dir);
            kb.append("owner", "repo-a", "comment a", "finding a");
            kb.append("owner", "repo-b", "comment b", "finding b");

            assertThat(kb.load("owner", "repo-a")).contains("finding a");
            assertThat(kb.load("owner", "repo-a")).doesNotContain("finding b");
            assertThat(kb.load("owner", "repo-b")).contains("finding b");
        }
    }

    @Nested
    class FileFor {

        @Test
        void pathTraversal_throwsSecurityException(@TempDir File dir) {
            PatternKnowledgeBase kb = new PatternKnowledgeBase(dir);
            org.assertj.core.api.Assertions.assertThatThrownBy(
                            () -> kb.append("../evil", "repo", "ctx", "response"))
                    .isInstanceOf(SecurityException.class);
        }

        @Test
        void normalOwnerRepo_doesNotThrow(@TempDir File dir) {
            PatternKnowledgeBase kb = new PatternKnowledgeBase(dir);
            // Should not throw for well-formed owner/repo names
            kb.append("owner", "repo", "ctx", "response");
        }
    }

    @Nested
    class Append {

        @Test
        void createsFileAndParentDirs(@TempDir File dir) {
            File nested = new File(dir, "sub/patterns");
            PatternKnowledgeBase kb = new PatternKnowledgeBase(nested);
            kb.append("owner", "repo", "comment", "finding");
            assertThat(kb.load("owner", "repo")).isNotBlank();
        }

        @Test
        void accumulatesMultipleEntries(@TempDir File dir) {
            PatternKnowledgeBase kb = new PatternKnowledgeBase(dir);
            kb.append("owner", "repo", "comment one", "finding one");
            kb.append("owner", "repo", "comment two", "finding two");

            String contents = kb.load("owner", "repo");
            assertThat(contents).contains("finding one");
            assertThat(contents).contains("finding two");
        }
    }

    @Nested
    class FormatEntry {

        @Test
        void containsQuotedContext() {
            String entry = PatternKnowledgeBase.formatEntry("the comment", "the finding");
            assertThat(entry).contains("> the comment");
        }

        @Test
        void containsResponse() {
            String entry = PatternKnowledgeBase.formatEntry("q", "the finding");
            assertThat(entry).contains("the finding");
        }

        @Test
        void multiLineContext_quotedPerLine() {
            String entry = PatternKnowledgeBase.formatEntry("line one\nline two", "finding");
            assertThat(entry).contains("> line one\n> line two");
        }

        @Test
        void containsDate() {
            String entry = PatternKnowledgeBase.formatEntry("q", "r");
            // ISO date like 2025-01-15 — just check the format
            assertThat(entry).containsPattern("\\d{4}-\\d{2}-\\d{2}");
        }

        @Test
        void startsWithSeparator() {
            String entry = PatternKnowledgeBase.formatEntry("q", "r");
            assertThat(entry).startsWith("\n---\n");
        }
    }
}

File: core/src/jvmTest/kotlin/com/jinloes/prpilot/services/PatternKnowledgeBaseTest.kt
package com.jinloes.prpilot.services

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldNotContain
import io.kotest.matchers.string.shouldStartWith
import io.kotest.matchers.string.shouldBeEmpty
import io.kotest.matchers.string.shouldNotBeEmpty
import io.kotest.matchers.string.shouldMatch
import java.io.File
import java.nio.file.Files

class PatternKnowledgeBaseTest : FunSpec({

    lateinit var tempDir: File
    beforeTest { tempDir = Files.createTempDirectory("pkb-test-").toFile() }
    afterTest { tempDir.deleteRecursively() }

    // ── load ─────────────────────────────────────────────────────────────

    context("load") {

        test("returns empty string when no file exists") {
            val kb = PatternKnowledgeBase(tempDir)
            kb.load("owner", "repo").shouldBeEmpty()
        }

        test("returns contents after append") {
            val kb = PatternKnowledgeBase(tempDir)
            kb.append("owner", "repo", "the comment", "the finding")
            kb.load("owner", "repo").shouldNotBeEmpty()
        }

        test("separates repos by file") {
            val kb = PatternKnowledgeBase(tempDir)
            kb.append("owner", "repo-a", "comment a", "finding a")
            kb.append("owner", "repo-b", "comment b", "finding b")

            kb.load("owner", "repo-a") shouldContain "finding a"
            kb.load("owner", "repo-a") shouldNotContain "finding b"
            kb.load("owner", "repo-b") shouldContain "finding b"
        }
    }

    // ── fileFor (via append) ─────────────────────────────────────────────

    context("fileFor") {

        test("path traversal in owner throws SecurityException") {
            val kb = PatternKnowledgeBase(tempDir)
            shouldThrow<SecurityException> {
                kb.append("../evil", "repo", "ctx", "response")
            }
        }

        test("normal owner/repo does not throw") {
            val kb = PatternKnowledgeBase(tempDir)
            kb.append("owner", "repo", "ctx", "response") // should not throw
        }
    }

    // ── append ───────────────────────────────────────────────────────────

    context("append") {

        test("creates file and parent directories") {
            val nested = File(tempDir, "sub/patterns")
            val kb = PatternKnowledgeBase(nested)
            kb.append("owner", "repo", "comment", "finding")
            kb.load("owner", "repo").shouldNotBeEmpty()
        }

        test("accumulates multiple entries in the same file") {
            val kb = PatternKnowledgeBase(tempDir)
            kb.append("owner", "repo", "comment one", "finding one")
            kb.append("owner", "repo", "comment two", "finding two")

            val contents = kb.load("owner", "repo")
            contents shouldContain "finding one"
            contents shouldContain "finding two"
        }
    }

    // ── formatEntry ──────────────────────────────────────────────────────

    context("formatEntry") {

        test("contains quoted context") {
            val entry = PatternKnowledgeBase.formatEntry("the comment", "the finding")
            entry shouldContain "> the comment"
        }

        test("contains response") {
            val entry = PatternKnowledgeBase.formatEntry("q", "the finding")
            entry shouldContain "the finding"
        }

        test("multi-line context is quoted per line") {
            val entry = PatternKnowledgeBase.formatEntry("line one\nline two", "finding")
            entry shouldContain "> line one\n> line two"
        }

        test("contains ISO date pattern") {
            val entry = PatternKnowledgeBase.formatEntry("q", "r")
            entry shouldMatch Regex("(?s).*\\d{4}-\\d{2}-\\d{2}.*")
        }

        test("starts with separator") {
            val entry = PatternKnowledgeBase.formatEntry("q", "r")
            entry shouldStartWith "\n---\n"
        }
    }
})
